### PR TITLE
Trello Custom Resource Definitions

### PR DESCRIPTION
**Pull request by mburke5678**
https://trello.com/c/dUBTcfhE/606-document-control-plane-custom-resource-definitions

### CODE REVIEW

**Comment by mburke5678** [comment]
@deads2k PTAL

**Comment by deads2k** [comment]
Straightforward copy of the kube docs?   Looks really similar

**Comment by mburke5678** [comment]
@deads2k Yes, this is an import and edit of the upstream doc. Is that appropriate in this case? Any difference in functionality between upstream and OpenShift proper?

**Comment by deads2k** [comment]
> @deads2k Yes, this is an import and edit of the upstream doc. Is that appropriate in this case? Any difference in functionality between upstream and OpenShift proper?

No difference.  It just seems like it will be a pain to maintain over time.

/lgtm

**Comment by mburke5678** [comment]
@wanghaoran1988 PTAL

**Comment by wanghaoran1988** [comment]
@deads2k @mburke5678 Generially LGFM, but as a user, I am wonder how can I write a controller for this specific Custom Resource, and how can I write a finalizer controller, do you think we should and this part docs ?

**Comment by deads2k** [comment]
> @deads2k @mburke5678 Generially LGFM, but as a user, I am wonder how can I write a controller for this specific Custom Resource, and how can I write a finalizer controller, do you think we should and this part docs ?

They don't exist upstream and we're still working on the experience (@sttts is working on making generation work for CRD types).  So, yes, but I don't think we should do it in 3.7.

File: admin_guide/custom_resource_definitions.adoc [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+----
+apiVersion: "stable.example.com/v1" <1>
+kind: CronTab <2>
+metadata:
+  name: my-new-cron-object <3>
+spec: <4>
+  cronSpec: "* * * * /5"
+  image: my-awesome-cron-image
+----
+
```
Comment (mburke5678) [inline admin_guide/custom_resource_definitions.adoc]: @deads2k @wanghaoran1988 What can go in the `spec` fields?

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL

File: _topic_map.yml [hunk 2]
```diff
@@ -575,6 +575,9 @@ Topics:
 - Name: Pruning Objects
   File: pruning_resources
   Distros: openshift-*
+- Name: Extending the Kubernets API with Custom Resources
```
Comment (vikram-redhat) [inline _topic_map.yml]: typo: should be: kubernetes

File: admin_guide/custom_resource_definitions.adoc [hunk 3]
```diff
@@ -0,0 +1,212 @@
+[[admin-guide-custom-resources]]
+= Extending the Kubernets API with Custom Resources
```
Comment (vikram-redhat) [inline admin_guide/custom_resource_definitions.adoc]: same typo for kubernetes.

File: architecture/additional_concepts/other_api_objects.adoc [hunk 4]
```diff
@@ -85,6 +85,13 @@ https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persis
 documentation] for more information.
 endif::[]
 
+=== Custom Resources
+
+A _custom resource_ is an extension of the Kubernetes API that extends the API or allows you to
+introduce your own API into a project or a cluster.
+
+See xref:../admin_guide/custom_resource_definitions.adoc#admin-guide-custom-resource[Extend the Kubernets API with Custom Resources].
```
Comment (vikram-redhat) [inline architecture/additional_concepts/other_api_objects.adoc]: typo for kubernetes

**Review by vikram-redhat — CHANGES_REQUESTED** [review]

File: _topic_map.yml (hunk 2 above)
Comment (mburke5678) [reply _topic_map.yml]: Thanks. All the times I have typed this word, odd that I got it wrong 3 times in the same PR!

File: admin_guide/custom_resource_definitions.adoc (hunk 1 above)
Comment (wanghaoran1988) [reply admin_guide/custom_resource_definitions.adoc]: IIUC, spec can be whatever you needed and just follow the yaml/json format, so controller can marshal/unmarshal to your defined structs.

**Comment by mburke5678** [comment]
[rev_history]
|xref:../admin_guide/custom_resource_definitions.adoc#admin-guide-custom-resources[Extending the Kubernetes API with Custom Resources]
|New topic on using custom resources to extend the Kubernetes API or introduce a new API.
%

File: _topic_map.yml [hunk 5]
```diff
@@ -575,6 +575,9 @@ Topics:
 - Name: Pruning Objects
   File: pruning_resources
   Distros: openshift-*
+- Name: Extending the Kubernetes API with Custom Resources
+  File: custom_resource_definitions
+  Distros: openshift-*
```
Comment (adellape) [inline _topic_map.yml]: @mburke5678 Do we know if this is relevant for OSD?

Comment (vikram-redhat) [reply _topic_map.yml]: @mburke5678 @adellape - I am not sure that this is relevant to OSD or Online. Michael has marked the label with enterprise only but the topic map makes it for all. I am going to remove it from dedicated and online till we know for sure.

**Comment by mburke5678** [comment]
@derekwaynecarr  @liggitt  @deads2k @wanghaoran1988  Is this feature in OpenShift Dedicated?

**Comment by deads2k** [comment]
> @derekwaynecarr @liggitt @deads2k @wanghaoran1988 Is this feature in OpenShift Dedicated?

It's present, but I don't know if dedicated admins have access.  I suspect not.  @abhgupta CRUD on `customresourcedefinitions.apiextensions.k8s.io`

**Comment by abhgupta** [comment]
No, dedicated admins cannot create custom resources

**Pull request merged**